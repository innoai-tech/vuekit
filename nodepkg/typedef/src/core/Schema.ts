import { isArray, isObject, isUndefined } from "es-toolkit/compat";
import type { OmitBy, PickBy } from "./Type.ts";

export type Simplify<T> = T extends any[] ? T : { [K in keyof T]: T[K] } & {};

export type Optionalize<S extends object> = OmitBy<S, undefined> &
  Partial<PickBy<S, undefined>>;

export type PropertyNames<O extends {}> = {
  [K in keyof O as K extends symbol ? never : K]: O[K];
};

export type Constructor = {
  new (...args: any[]): any;
};

export class Schema {
  static RecordKey = Symbol("RecordKey");

  static meta = Symbol("meta");
  static optional = Symbol("optional");
  static underlying = Symbol("underlying");
  static unwrap = Symbol("unwrap");

  static extractSchema = (v: any) => {
    return new Collector().toValue(v);
  };

  static extractMeta = (v: any) => {
    return new Collector().toValue(v["meta"] ?? v[Schema.meta]);
  };

  static create<S extends {}>(
    schema: S,
  ): Simplify<
    PropertyNames<S> & {
      metadata: Record<string, any>;
    }
  >;
  static create<S extends {}, E extends {}>(
    schema: S,
    origin: E,
  ): Simplify<PropertyNames<S> & PropertyNames<E>>;
  static create<S extends {}, E extends {}>(
    schema: S,
    origin: E,
    scope?: symbol,
  ): Simplify<PropertyNames<S> & PropertyNames<E>>;
  static create<S extends {}, E extends {}>(
    base: S,
    origin?: E,
    scope?: symbol,
  ): Simplify<PropertyNames<S> & PropertyNames<E>> {
    const parent = origin ?? (base as any)[Schema.underlying]?.["schema"] ?? {};

    let meta: Schema | undefined;

    return new Proxy(
      {},
      {
        ownKeys(_: {}): ArrayLike<string | symbol> {
          const m = new Map<string, string>();

          if (scope) {
            if ((base as any)[scope]) {
              for (const x of Object.getOwnPropertyNames(
                (base as any)[scope],
              )) {
                m.set(x, x);
              }
            }

            if (parent) {
              if ((parent as any)[scope]) {
                for (const x of Object.getOwnPropertyNames(
                  (parent as any)[scope],
                )) {
                  m.set(x, x);
                }
              }
            }
          } else {
            for (const x of Object.getOwnPropertyNames(base)) {
              m.set(x, x);
            }

            if (parent) {
              for (const x of Object.getOwnPropertyNames(parent)) {
                m.set(x, x);
              }
            }
          }

          return [...m.keys()];
        },
        get(_: {}, p: string | symbol): any {
          if (p === Schema.meta) {
            return (meta ??= Schema.create(base, parent, Schema.meta) as any);
          }

          if (scope) {
            if (isObject(base)) {
              const v = (base as any)?.[scope]?.[p];
              if (!isUndefined(v)) {
                return v;
              }
            }
            return parent ? (parent as any)?.[scope]?.[p] : undefined;
          }

          if (isObject(base)) {
            const v = (base as any)?.[p];
            if (!isUndefined(v)) {
              return v;
            }
          }

          return parent ? (parent as any)[p] : undefined;
        },
      },
    ) as any;
  }

  static schemaProp<T = any>(
    withSchema: { schema: any },
    key: string | symbol,
  ): T | undefined {
    return getSchema(withSchema.schema, key);
  }

  static metaProp<T = any>(
    withSchema: { schema: any },
    key: string | symbol,
  ): T | undefined {
    return getMeta(withSchema.schema, key);
  }
}

const getSchema = (schema: any, key: any): any => {
  if (!isObject(schema)) {
    return undefined;
  }
  return (
    (schema as any)[key] ??
    ((schema as any)[Schema.unwrap]
      ? getSchema((schema as any)[Schema.unwrap]().schema, key)
      : undefined)
  );
};

const getMeta = (schema: any, metaKey: any): any => {
  if (!isObject(schema)) {
    return undefined;
  }
  return (
    (schema as any)[Schema.meta]?.[metaKey] ??
    ((schema as any)[Schema.unwrap]
      ? getMeta((schema as any)[Schema.unwrap]().schema, metaKey)
      : undefined)
  );
};

class Collector {
  constructor(private deref: boolean = false) {}

  toValue(v: any): any {
    if (v) {
      if (!isUndefined(v["schema"])) {
        return this.toValue(v["schema"]);
      }

      if (this.deref) {
        if (v[Schema.unwrap]) {
          return this.toValue(v[Schema.unwrap]());
        }
      }

      if (isArray(v)) {
        return this.toArray(v);
      }

      if (v.constructor === Object) {
        return this.toObject(v);
      }
    }

    return v;
  }

  toObject(s: any): any {
    const x: Record<string, any> = {};

    for (const k of Object.getOwnPropertyNames(s)) {
      x[k] = this.toValue(s[k]);
    }
    return x;
  }

  toArray(values: any[]): any[] {
    const arr: any[] = [];
    for (const value of values) {
      arr.push(this.toValue(value));
    }
    return arr;
  }
}
