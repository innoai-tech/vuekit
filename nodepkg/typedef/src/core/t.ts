import {TypeRef} from "./TypeRef.ts";
import {TypeAny} from "./TypeAny.ts";
import {DefaultedType, OptionalType, Type, TypeNever, TypeWrapper} from "./Type.ts";
import {TypeNull} from "./TypeNull.ts";
import {TypeString} from "./TypeString.ts";
import {TypeNumber} from "./TypeNumber.ts";
import {TypeInteger} from "./TypeInteger.ts";
import {TypeBoolean} from "./TypeBoolean.ts";
import {TypeBinary} from "./TypeBinary.ts";
import {TypeEnum} from "./TypeEnum.ts";
import {TypeObject} from "./TypeObject.ts";
import {TypeRecord} from "./TypeRecord.ts";
import {TypeArray} from "./TypeArray.ts";
import {TypeTuple} from "./TypeTuple.ts";
import {TypeIntersection} from "./TypeIntersection.ts";
import {TypeUnion} from "./TypeUnion.ts";

export const ref = TypeRef.create;
export const any = TypeAny.create;
export const never = TypeNever.create;
export const nil = TypeNull.create;

export const string = TypeString.create;

export function pattern(pattern: RegExp, msg?: string) {
    return <T extends string, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (pattern.test(value)) {
                return true
            }
            return msg ?? `Expected a ${type.type} matching \`/${pattern.source}/\` but received "${value}"`
        }, {
            pattern,
        })
    };
}

export const number = TypeNumber.create;
export const integer = TypeInteger.create;

export function minimum<M extends number>(min: M, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value >= min) {
                return true
            }
            return msg ?? `Expected value great than or equal ${min}, but received "${value}"`
        }, {
            minimum: min,
        })
    };
}

export function exclusiveMinimum<M extends number>(min: M, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value > min) {
                return true
            }
            return (msg ?? `Expected value great than ${min}, but received "${value}"`)
        }, {
            exclusiveMinimum: min,
        })
    };
}


export function maximum(max: number, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value <= max) {
                return true
            }
            return (msg ?? `Expected value less than or equal ${max}, but received "${value}"`)
        }, {
            maximum: max,
        }) as any
    };
}


export function exclusiveMaximum(max: number, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value < max) {
                return true
            }
            return msg ?? `Expected value less than or equal ${max}, but received "${value}"`
        }, {
            exclusiveMaximum: max,
        }) as any;
    };
}


export function multipleOf(multipleOf: number, msg?: string) {
    return <T extends number, S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (value % multipleOf == 0) {
                return true
            }
            return msg ?? `Expected value multiple of ${multipleOf}, but received "${value}"`
        }, {
            multipleOf: multipleOf,
        });
    };
}

export const boolean = TypeBoolean.create;
export const binary = TypeBinary.create;

export const enums = TypeEnum.create;
export const nativeEnum = TypeEnum.nativeEnum;
export const literal = TypeEnum.literal;

export const object = TypeObject.create;
export const record = TypeRecord.create;

export const tuple = TypeTuple.create;
export const array = TypeArray.create;

export function minItems(minItems: number, msg?: string) {
    return <T extends any[], S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (Array.isArray(value) && value.length >= minItems) {
                return true
            }
            return msg ?? `Expected array value at least ${minItems}, but received "${value?.length}"`
        }, {
            minItems: minItems,
        })
    };
}

export function maxItems(maxItems: number, msg?: string) {
    return <T extends any[], S>(type: Type<T, S>) => {
        return TypeWrapper.refine(type, (value) => {
            if (Array.isArray(value) && value.length <= maxItems) {
                return true
            }
            return msg ?? `Expected array value  ${maxItems}, but received "${value?.length}"`
        }, {
            maxItems: maxItems,
        });
    };
}

export const intersection = TypeIntersection.create;
export const union = TypeUnion.create;
export const discriminatorMapping = TypeUnion.discriminatorMapping;

export const custom = Type.define;
export const refine = TypeWrapper.refine;

export function defaults<T>(v: T) {
    return <U extends Type<T, any>>(type: U): U => {
        return DefaultedType.create(type, v) as unknown as U;
    };
}

export function optional() {
    return <T, S>(type: Type<T, S>): Type<T | undefined, S> => {
        return OptionalType.create(type) as unknown as any;
    };
}

export function annotate<M extends Record<string, any>>(meta: M) {
    return <T, S>(type: Type<T, S>): Type<T, S & M> => {
        return TypeWrapper.of(type, {$meta: meta}) as unknown as Type<T, S & M>;
    };
}
