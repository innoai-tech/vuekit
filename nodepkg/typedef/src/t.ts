export * from "./types";
export * from "./modifers";

