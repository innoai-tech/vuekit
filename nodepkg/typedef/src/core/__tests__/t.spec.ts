import { describe, expect, test } from "vitest";
import * as t from "../t.ts";
import { Schema } from "../Schema.ts";

enum Gender {
  MALE = "MALE",
  FEMALE = "FEMALE",
}

describe("define type", () => {
  test("should combine metadata", () => {
    const s = t
      .string()
      .use(t.annotate({ label: "label" }), t.annotate({ title: "title" }))
      .optional();

    expect(s.schema.type).toEqual("string");
    expect(s.meta["label"]).toEqual("label");
    expect(s.meta["title"]).toEqual("title");

    const ref = t.ref("S", () => s);

    expect(Schema.schemaProp<string>(ref, "type")).toEqual("string");
    expect(Schema.metaProp<string>(ref, "label")).toEqual("label");
    expect(Schema.metaProp<string>(ref, "title")).toEqual("title");
  });

  describe("should create string", () => {
    const s = t.string();

    test("should validate", () => {
      expect(Schema.extractSchema(s)).toEqual({
        type: "string",
      });

      expect(s.validate(1)[0]).not.toBeUndefined();
      expect(s.validate("")[0]).toBeUndefined();
    });

    test("with default", () => {
      const s2 = s.default("1");

      expect(s2.create(undefined)).toBe("1");
    });

    test("with optional", () => {
      const s2 = s.optional();

      expect(Schema.extractSchema(s2)).toEqual({
        type: "string",
      });

      expect(s2.validate(undefined)[0]).toBeUndefined();
      expect(s2.validate("")[0]).toBeUndefined();
      expect(s2.validate(1)[0]).not.toBeUndefined();
    });

    test("with modifiers", () => {
      const s2 = s.use(t.minLength(1), t.maxLength(10));

      expect(Schema.extractSchema(s2.schema)).toEqual({
        type: "string",
        minLength: 1,
        maxLength: 10,
      });

      expect(s2.validate("")[0]).not.toBeUndefined();
    });
  });

  class A {
    type = "a" as const;
  }

  class B {
    type = "b" as const;

    @t.string()
    name!: string;
  }

  const types = [
    t.any(),
    t.never(),
    t.nil(),
    t.binary(),
    t.boolean(),
    t.integer(),
    t.number(),
    t.enums(["1", "2", "3"]),
    t.nativeEnum(Gender),
    t.record(t.string(), t.number()),
    t.array(t.string()),
    t.tuple([t.number(), t.number()]),
    t.intersection(
      t.ref("X", () => {
        return t.object({
          name: t.string(),
        });
      }),
      t.object({
        age: t.integer().optional(),
      }),
    ),
    t.object({
      name: t.string(),
      age: t.integer().optional(),
    }),
    t.union(t.integer(), t.string()),
    t.discriminatorMapping("type", {
      a: t.object(),
      b: t.object({
        name: t.string(),
      }),
    }),
    t.discriminatorMapping("type", A, B),
  ];

  for (const s of types) {
    test(s.type, () => {
      expect(Schema.extractSchema(s)).toMatchSnapshot();
    });
  }
});

describe("define type with metadata", () => {
  test("create object", () => {
    class Account {
      @t.string()
      @t.annotate({ label: "姓名" })
      @t.annotate({ readOnlyWhenExists: true })
      name!: string;

      @t.nativeEnum(Gender)
      @t.optional()
      @t.annotate({ label: "性别" })
      gender?: Gender;

      @t.integer()
      @t.annotate({ label: "年龄" })
      age!: number;
    }

    const s = t.object(Account);

    expect(Schema.extractSchema(s.schema)).toEqual({
      type: "object",
      properties: {
        name: {
          type: "string",
        },
        gender: {
          enum: ["MALE", "FEMALE"],
        },
        age: {
          type: "integer",
        },
      },
      required: ["name", "age"],
      additionalProperties: false,
    });

    const metadata: Record<string, any> = {};

    for (const [k, _, prop] of s.entries({})) {
      if (prop.type === "never") {
        continue;
      }
      metadata[k as string] = Schema.extractSchema(prop.meta);
    }

    expect(metadata).toEqual({
      name: {
        label: "姓名",
        readOnlyWhenExists: true,
      },
      gender: {
        label: "性别",
      },
      age: {
        label: "年龄",
      },
    });
  });

  describe("create tagged union", () => {
    class A {
      @t.annotate({ label: "类型" })
      type = "a" as const;
    }

    class B {
      @t.annotate({ label: "类型" })
      type = "b" as const;

      @t.string()
      @t.annotate({ label: "姓名" })
      name!: string;
    }

    const s = t.discriminatorMapping("type", A, B);

    test("when discriminator not set", () => {
      const metadata: Record<string, any> = {};

      for (const [k, _, prop] of s.entries({})) {
        metadata[k as string] = Schema.extractSchema(prop.meta);
      }

      expect(metadata).toEqual({
        type: {
          label: "类型",
        },
      });
    });

    test("when discriminator type with a", () => {
      const metadata: Record<string, any> = {};

      for (const [k, _, prop] of s.entries({
        type: "a",
      })) {
        metadata[k as string] = Schema.extractSchema(prop.meta);
      }

      expect(metadata).toEqual({
        type: {
          label: "类型",
        },
      });
    });

    test("when discriminator type with b", () => {
      const metadata: Record<string, any> = {};

      for (const [k, _, prop] of s.entries({ type: "b" })) {
        metadata[k as string] = Schema.extractSchema(prop.meta);
      }

      expect(metadata).toEqual({
        type: {
          label: "类型",
        },
        name: {
          label: "姓名",
        },
      });
    });
  });
});
