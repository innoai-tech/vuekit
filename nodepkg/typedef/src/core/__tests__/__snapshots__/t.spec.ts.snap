// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`define type any 1`] = `{}`;

exports[`define type never 1`] = `false`;

exports[`define type null 1`] = `
{
  "type": "null",
}
`;

exports[`define type binary 1`] = `
{
  "format": "binary",
  "type": "string",
}
`;

exports[`define type boolean 1`] = `
{
  "type": "boolean",
}
`;

exports[`define type integer 1`] = `
{
  "type": "integer",
}
`;

exports[`define type number 1`] = `
{
  "type": "number",
}
`;

exports[`define type enums 1`] = `
{
  "enum": [
    "1",
    "2",
    "3",
  ],
}
`;

exports[`define type enums 2`] = `
{
  "enum": [
    "MALE",
    "FEMALE",
  ],
}
`;

exports[`define type record 1`] = `
{
  "additionalProperties": {
    "type": "number",
  },
  "propertyNames": {
    "type": "string",
  },
  "type": "object",
}
`;

exports[`define type array 1`] = `
{
  "items": {
    "type": "string",
  },
  "type": "array",
}
`;

exports[`define type tuple 1`] = `
{
  "items": [
    {
      "type": "number",
    },
    {
      "type": "number",
    },
  ],
  "type": "array",
}
`;

exports[`define type intersection 1`] = `
{
  "allOf": [
    {
      "$ref": "X",
    },
    {
      "additionalProperties": false,
      "properties": {
        "age": {
          "type": "integer",
        },
      },
      "required": [],
      "type": "object",
    },
  ],
}
`;

exports[`define type object 1`] = `
{
  "additionalProperties": false,
  "properties": {
    "age": {
      "type": "integer",
    },
    "name": {
      "type": "string",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;

exports[`define type union 1`] = `
{
  "oneOf": [
    {
      "type": "integer",
    },
    {
      "type": "string",
    },
  ],
}
`;

exports[`define type union 2`] = `
{
  "discriminator": {
    "propertyName": "type",
  },
  "oneOf": [
    {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "a",
          ],
        },
      },
      "required": [
        "type",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string",
        },
        "type": {
          "enum": [
            "b",
          ],
        },
      },
      "required": [
        "type",
        "name",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`define type union 3`] = `
{
  "discriminator": {
    "propertyName": "type",
  },
  "oneOf": [
    {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "a",
          ],
        },
      },
      "required": [
        "type",
      ],
      "type": "object",
    },
    {
      "additionalProperties": false,
      "properties": {
        "name": {
          "type": "string",
        },
        "type": {
          "enum": [
            "b",
          ],
        },
      },
      "required": [
        "type",
        "name",
      ],
      "type": "object",
    },
  ],
}
`;
